lib/galaxy/tools/evaluation.py

```diff
@@ -639,6 +639,24 @@ def _build_environment_variables(self):
                 else:
                     environment_variable_template = ""
                 is_template = False
+            elif inject == "oidc_refresh_token":
+                if self._user:
+                    environment_variable_template = self._user.oidc_refresh_token
```

Comment:
Will tools be using the refresh token? My understanding is that for a refresh token to be used to generate a new access token, the client id and secret must also be available.

Reply:
> 1. Should we consider adding an allowlist to gate token injection, similar to how sanitize_allowlist is used on the client side?

What is the scenario for a security breach? Someone installs a malicious tool in Galaxy? Or does a normal tool runs in an untrusted environment? Or both? In the first case, sanitize_allowlist might help. In the second  - it would probably be better to have an allowed list of destinations for a tool. But this would complicate administration, I guess. 

I saw that there are no restrictions to expose Galaxy API tokens, so I just added OIDC tokens in the same way. Actually, losing the  API token is worse than an OIDC one since that one is time unlimited. 

> 2. What are the implications of token expiry during tool execution, something that's more likely to happen in high-load sites where jobs may sit for a long time in the queue.

Agree, this is a weak point in the whole scenario. A tool that needs to use OIDC tokens should not sit in a queue for too long. This is also why we expose a refresh token. But then, as you've noticed, you need to know the client id and secrets. We use it for one tool, passing corresponding values from the Galaxy config file (https://github.com/PaNGalaxy/neutron-tools/blob/main/tools/neutrons/import_oncat.xml#L18). Btw, the fact that we can do that is another security issue where one can write a tool that grabs Galaxy config from a server!

> 
> I would like to see this feature being available especially in trusted environments, where the entire stack, inclusive of the job execution environment, tools themselves, and everything in between, is trusted and secure. However, I think it's a feature that may need some access controls in other environments where there's less confidence in the tools in particular.

Agree, we developed that with our environment in mind, where everything is trusted and is under our control. So, would adding sanitize_allowlist be good enough as a first step? At least with that, by default, the feature is not available, and then one can switch it on for specific tools.

Reply:
After thinking about this a bit more, I don't think my original concerns are relevant. While it would be nice to be able to enforce controls on specific tools, I think that a general permissions model for tools would be a far different undertaking and is not really within the scope of this PR. As you've pointed out, there are too many ways to bypass those things anyway, and we would probably need something like an oauth2 consent page at the point of installing the tool for a tool to request the specific permissions it needs and for admins to approve it. Therefore, I think what you've done seems ok to me.